code/modules/reagents/reagent_containers/hypospray.dm

```diff
@@ -60,7 +60,57 @@
 		used = !used
 
 	return
-
+//A vial-loaded hypospray. Cartridge-based!
+/obj/item/weapon/reagent_containers/hypospray/vial
+	name = "hypospray mkII"
+	desc = "A new development from DeForest Medical, this new hypospray takes 30-unit vials as the drug supply for easy swapping."
+	var/obj/item/weapon/reagent_containers/glass/beaker/vial/loaded_vial //Wow, what a name.
+	volume = 0
+
+/obj/item/weapon/reagent_containers/hypospray/vial/New()
+	..()
+	loaded_vial = new /obj/item/weapon/reagent_containers/glass/beaker/vial(src) //Comes with an empty vial
+	volume = loaded_vial.volume
+	reagents.maximum_volume = loaded_vial.reagents.maximum_volume
+
+/obj/item/weapon/reagent_containers/hypospray/vial/attack_hand(mob/user as mob)
+	if(user.get_inactive_hand() == src)
+		if(loaded_vial)
+			reagents.trans_to_holder(loaded_vial.reagents,volume)
+			reagents.maximum_volume = 0
+			loaded_vial.update_icon()
+			user.put_in_hands(loaded_vial)
+			loaded_vial = null
+			user << "<span class='notice'>You remove the vial from the [src].</span>"

```

Comment:
to_chat(user, "<span class='notice'>You remove the vial from the [src].</span>")